In [152]:
import pandas as pd
import numpy as np

In [153]:
"""#reading the file using dask
import dask
import dask.dataframe as dd
from dask.delayed import delayed

parts = dask.delayed(pd.read_excel)("N:/15.11.2023 Data/dataexport_20231115T124544.xlsx")
data = dd.from_delayed(parts)

data.head()"""

'#reading the file using dask\nimport dask\nimport dask.dataframe as dd\nfrom dask.delayed import delayed\n\nparts = dask.delayed(pd.read_excel)("N:/15.11.2023 Data/dataexport_20231115T124544.xlsx")\ndata = dd.from_delayed(parts)\n\ndata.head()'

### 1. Readind in the data

In [154]:
data = pd.read_excel('N:/15.11.2023 Data/dataexport_20231115T124544.xlsx')

C:\Users\phall\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\openpyxl\styles\stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


In [155]:
data.head()

location        Basel              Basel.1      Basel.2      Basel.3  \
0       lat        47.75                47.75        47.75        47.75   
1       lon          7.5                  7.5          7.5          7.5   
2       asl    363.65314            363.65314    363.65314    363.65314   
3  variable  Temperature  Growing Degree Days  Temperature  Temperature   
4      unit           °C                 GDDc           °C           °C   

       Basel.4      Basel.5      Basel.6              Basel.7  \
0        47.75        47.75        47.75                47.75   
1          7.5          7.5          7.5                  7.5   
2    363.65314    363.65314    363.65314            363.65314   
3  Temperature  Temperature  Temperature  Precipitation Total   
4           °C           °C           °C                   mm   

             Basel.8  ...          Basel.53          Basel.54       Basel.55  \
0              47.75  ...             47.75             47.75          47.75   
1                7.5  ...               7.5               7.5            7.5   
2          363.65314  ...         363.65314         363.65314      363.65314   
3  Relative Humidity  ...  Soil Temperature  Soil Temperature  Soil Moisture   
4                  %  ...                °C                °C          m³/m³   

        Basel.56       Basel.57       Basel.58  \
0          47.75          47.75          47.75   
1            7.5            7.5            7.5   
2      363.65314      363.65314      363.65314   
3  Soil Moisture  Soil Moisture  Soil Moisture   
4          m³/m³          m³/m³          m³/m³   

                           Basel.59                          Basel.60  \
0                             47.75                             47.75   
1                               7.5                               7.5   
2                         363.65314                         363.65314   
3  Soil Moisture Available To Plant  Soil Moisture Available To Plant   
4                          fraction                          fraction   

                           Basel.61                          Basel.62  
0                             47.75                             47.75  
1                               7.5                               7.5  
2                         363.65314                         363.65314  
3  Soil Moisture Available To Plant  Soil Moisture Available To Plant  
4                          fraction                          fraction  

[5 rows x 64 columns]

### 2. Data Cleaning

First 4 rows(3 plus current heading) are the same thing which can be replaced

The data is from Basel(location) found at latitude=47.75 and longitude=7.5. Resolution is measured hourly with no aggregation

In [156]:
data = data.iloc[4:]

In [157]:
data = data.drop([5,7])
data.head()

location                                        Basel  \
4                  unit                                           °C   
6            resolution                                       hourly   
8             timestamp  Basel Temperature [2 m elevation corrected]   
9   2000-01-01 00:00:00                                     1.550245   
10  2000-01-01 01:00:00                                     1.500245   

                                              Basel.1  \
4                                                GDDc   
6                                              hourly   
8   Basel Growing Degree Days [2 m elevation corre...   
9                                                   0   
10                                                  0   

                       Basel.2                     Basel.3  \
4                           °C                          °C   
6                       hourly                      hourly   
8   Basel Temperature [900 mb]  Basel Temperature [850 mb]   
9                        -0.34                       -2.35   
10                        -0.3                       -2.63   

                       Basel.4                     Basel.5  \
4                           °C                          °C   
6                       hourly                      hourly   
8   Basel Temperature [800 mb]  Basel Temperature [700 mb]   
9                        -4.99                       -8.75   
10                       -4.67                       -8.95   

                       Basel.6                    Basel.7  \
4                           °C                         mm   
6                       hourly                     hourly   
8   Basel Temperature [500 mb]  Basel Precipitation Total   
9                       -23.01                          0   
10                      -22.71                          0   

                          Basel.8  ...  \
4                               %  ...   
6                          hourly  ...   
8   Basel Relative Humidity [2 m]  ...   
9                        93.43871  ...   
10                       93.29928  ...   

                                   Basel.53  \
4                                        °C   
6                                    hourly   
8   Basel Soil Temperature [28-100 cm down]   
9                                      3.85   
10                                     3.81   

                                    Basel.54  \
4                                         °C   
6                                     hourly   
8   Basel Soil Temperature [100-255 cm down]   
9                                       7.54   
10                                      7.63   

                             Basel.55                            Basel.56  \
4                               m³/m³                               m³/m³   
6                              hourly                              hourly   
8   Basel Soil Moisture [0-7 cm down]  Basel Soil Moisture [7-28 cm down]   
9                               0.382                               0.385   
10                              0.384                               0.386   

                                Basel.57  \
4                                  m³/m³   
6                                 hourly   
8   Basel Soil Moisture [28-100 cm down]   
9                                  0.416   
10                                 0.418   

                                 Basel.58  \
4                                   m³/m³   
6                                  hourly   
8   Basel Soil Moisture [100-255 cm down]   
9                                   0.395   
10                                  0.401   

                                             Basel.59  \
4                                            fraction   
6                                              hourly   
8   Basel Soil Moisture Available To Plant [0-7 cm...   
9                                               0.231   
10               

In [158]:
data.columns = data.iloc[2]
data = data.drop(8)
data = data.drop(6)

# After dropping the necessary rows, now we can rename the columns accordingly with their unit of measurent
# Drop the 'Basel ' prefix from all column names
data.columns = data.columns.str.replace('Basel ', '')

# Concatenate column names with row 1 values in brackets
data.columns = [f'{col} ({data.iloc[0][col]})' for col in data.columns]

# Drop the first row, as it's now included in the column names
data = data.iloc[1:]

# Now we reset the index of the dataframe
data.reset_index()

data.head()

timestamp (unit) Temperature [2 m elevation corrected] (°C)  \
9   2000-01-01 00:00:00                                   1.550245   
10  2000-01-01 01:00:00                                   1.500245   
11  2000-01-01 02:00:00                                   1.470245   
12  2000-01-01 03:00:00                                   1.880245   
13  2000-01-01 04:00:00                                   2.310246   

   Growing Degree Days [2 m elevation corrected] (GDDc)  \
9                                                   0     
10                                                  0     
11                                                  0     
12                                                  0     
13                                                  0     

   Temperature [900 mb] (°C) Temperature [850 mb] (°C)  \
9                      -0.34                     -2.35   
10                      -0.3                     -2.63   
11                     -0.18                     -2.75   
12                     -0.15                     -2.77   
13                     -0.29                     -2.74   

   Temperature [800 mb] (°C) Temperature [700 mb] (°C)  \
9                      -4.99                     -8.75   
10                     -4.67                     -8.95   
11                     -4.52                     -8.63   
12                     -4.37                        -8   
13                     -4.44                     -7.59   

   Temperature [500 mb] (°C) Precipitation Total (mm)  \
9                     -23.01                        0   
10                    -22.71                        0   
11                    -22.79                        0   
12                    -22.85                        0   
13                    -22.62                        0   

   Relative Humidity [2 m] (%)  ... Soil Temperature [28-100 cm down] (°C)  \
9                     93.43871  ...                                   3.85   
10                    93.29928  ...                                   3.81   
11                    93.43304  ...                                    3.8   
12                    93.80152  ...                                    3.8   
13                    93.96607  ...                                   3.79   

   Soil Temperature [100-255 cm down] (°C)  \
9                                     7.54   
10                                    7.63   
11                                    7.63   
12                                    7.63   
13                                    7.62   

   Soil Moisture [0-7 cm down] (m³/m³) Soil Moisture [7-28 cm down] (m³/m³)  \
9                                0.382                                0.385   
10                               0.384                                0.386   
11                               0.384                                0.385   
12                               0.384                                0.385   
13                               0.384                                0.385   

   Soil Moisture [28-100 cm down] (m³/m³)  \
9                                   0.416   
10                                  0.418   
11                                  0.418   
12                                  0.418   
13                                  0.417   

   Soil Moisture [100-255 cm down] (m³/m³)  \
9                                    0.395   
10                                   0.401   
11                                   0.401   
12                                   0.401   
13                                   0.401   

   Soil Moisture Available To Plant [0-7 cm down] (fraction)  \
9                                               0.231          
10                                              0.233          
11                                              0.233          
12                                              0.233          
13                                              0.233          

   Soil Moisture Available To Plant 

In [159]:
data.columns

Index(['timestamp (unit)', 'Temperature [2 m elevation corrected] (°C)',
       'Growing Degree Days [2 m elevation corrected] (GDDc)',
       'Temperature [900 mb] (°C)', 'Temperature [850 mb] (°C)',
       'Temperature [800 mb] (°C)', 'Temperature [700 mb] (°C)',
       'Temperature [500 mb] (°C)', 'Precipitation Total (mm)',
       'Relative Humidity [2 m] (%)', 'Snowfall Amount (cm)', 'Snow Depth (m)',
       'Wind Gust (km/h)', 'Wind Speed [10 m] (km/h)',
       'Wind Direction [10 m] (°)', 'Wind Speed [100 m] (km/h)',
       'Wind Direction [100 m] (°)', 'Wind Speed [900 mb] (km/h)',
       'Wind Direction [900 mb] (°)', 'Wind Speed [850 mb] (km/h)',
       'Wind Direction [850 mb] (°)', 'Wind Speed [800 mb] (km/h)',
       'Wind Direction [800 mb] (°)', 'Wind Speed [700 mb] (km/h)',
       'Wind Direction [700 mb] (°)', 'Wind Speed [500 mb] (km/h)',
       'Wind Direction [500 mb] (°)', 'Wind Speed [250 mb] (km/h)',
       'Wind Direction [250 mb] (°)', 'Cloud Cover Total (%)',


In [160]:
# Rename the first column of the dataset from 'Timestamp Unit' to just 'Timestamp'
data = data.rename(columns={"timestamp (unit)": "Timestamp"})

In [161]:
data.tail()

Timestamp Temperature [2 m elevation corrected] (°C)  \
209284  2023-11-15 19:00:00                                        NaN   
209285  2023-11-15 20:00:00                                        NaN   
209286  2023-11-15 21:00:00                                        NaN   
209287  2023-11-15 22:00:00                                        NaN   
209288  2023-11-15 23:00:00                                        NaN   

       Growing Degree Days [2 m elevation corrected] (GDDc)  \
209284                                                NaN     
209285                                                NaN     
209286                                                NaN     
209287                                                NaN     
209288                                                NaN     

       Temperature [900 mb] (°C) Temperature [850 mb] (°C)  \
209284                       NaN                       NaN   
209285                       NaN                       NaN   
209286                       NaN                       NaN   
209287                       NaN                       NaN   
209288                       NaN                       NaN   

       Temperature [800 mb] (°C) Temperature [700 mb] (°C)  \
209284                       NaN                       NaN   
209285                       NaN                       NaN   
209286                       NaN                       NaN   
209287                       NaN                       NaN   
209288                       NaN                       NaN   

       Temperature [500 mb] (°C) Precipitation Total (mm)  \
209284                       NaN                      NaN   
209285                       NaN                      NaN   
209286                       NaN                      NaN   
209287                       NaN                      NaN   
209288                       NaN                      NaN   

       Relative Humidity [2 m] (%)  ...  \
209284                         NaN  ...   
209285                         NaN  ...   
209286                         NaN  ...   
209287                         NaN  ...   
209288                         NaN  ...   

       Soil Temperature [28-100 cm down] (°C)  \
209284                                    NaN   
209285                                    NaN   
209286                                    NaN   
209287                                    NaN   
209288                                    NaN   

       Soil Temperature [100-255 cm down] (°C)  \
209284                                     NaN   
209285                                     NaN   
209286                                     NaN   
209287                                     NaN   
209288                                     NaN   

       Soil Moisture [0-7 cm down] (m³/m³)  \
209284                                 NaN   
209285                                 NaN   
209286                                 NaN   
209287                                 NaN   
209288                                 NaN   

       Soil Moisture [7-28 cm down] (m³/m³)  \
209284                                  NaN   
209285                                  NaN   
209286                                  NaN   
209287                                  NaN   
209288                                  NaN   

       Soil Moisture [28-100 cm down] (m³/m³)  \
209284                                    NaN   
209285                                    NaN   
209286                                    NaN   
209287                                    NaN   
209288                                    NaN   

       Soil Moisture [100-255 cm down] (m³/m³)  \
209284                                     NaN   
209285                                     NaN   
209286                                     NaN   
209287                                     NaN   
209288                                     NaN   

       Soil Moisture Available To Plant [0-7 cm down] (fraction)  \
20

In [162]:
data.shape

(209280, 64)

In [163]:
data.isnull().sum()

Timestamp                                                          0
Temperature [2 m elevation corrected] (°C)                       191
Growing Degree Days [2 m elevation corrected] (GDDc)             191
Temperature [900 mb] (°C)                                        191
Temperature [850 mb] (°C)                                        191
                                                                ... 
Soil Moisture [100-255 cm down] (m³/m³)                          191
Soil Moisture Available To Plant [0-7 cm down] (fraction)        191
Soil Moisture Available To Plant [7-28 cm down] (fraction)       191
Soil Moisture Available To Plant [28-100 cm down] (fraction)     191
Soil Moisture Available To Plant [100-255 cm down] (fraction)    191
Length: 64, dtype: int64

In [164]:
# 209280-191 = 209089
print(data.iloc[209088])

Timestamp                                                        2023-11-08 00:00:00
Temperature [2 m elevation corrected] (°C)                                  5.160245
Growing Degree Days [2 m elevation corrected] (GDDc)                               0
Temperature [900 mb] (°C)                                                       2.84
Temperature [850 mb] (°C)                                                      -0.26
                                                                        ...         
Soil Moisture [100-255 cm down] (m³/m³)                                        0.301
Soil Moisture Available To Plant [0-7 cm down] (fraction)                      0.227
Soil Moisture Available To Plant [7-28 cm down] (fraction)                     0.222
Soil Moisture Available To Plant [28-100 cm down] (fraction)                   0.127
Soil Moisture Available To Plant [100-255 cm down] (fraction)                   0.15
Name: 209097, Length: 64, dtype: object


This suggests that the last 191 tuples are just null valued, so we drop the last 191 rows

In [165]:
data = data.iloc[:-191]
data.tail()

Timestamp Temperature [2 m elevation corrected] (°C)  \
209093  2023-11-07 20:00:00                                   6.510245   
209094  2023-11-07 21:00:00                                   5.970245   
209095  2023-11-07 22:00:00                                   5.700245   
209096  2023-11-07 23:00:00                                   5.450245   
209097  2023-11-08 00:00:00                                   5.160245   

       Growing Degree Days [2 m elevation corrected] (GDDc)  \
209093                                                  0     
209094                                                  0     
209095                                                  0     
209096                                                  0     
209097                                                  0     

       Temperature [900 mb] (°C) Temperature [850 mb] (°C)  \
209093                      3.29                      0.06   
209094                      3.32                      0.02   
209095                      3.25                     -0.12   
209096                      3.35                     -0.21   
209097                      2.84                     -0.26   

       Temperature [800 mb] (°C) Temperature [700 mb] (°C)  \
209093                     -3.67                    -10.72   
209094                     -3.66                    -10.71   
209095                     -3.59                    -10.49   
209096                     -3.33                    -10.01   
209097                     -3.53                    -10.65   

       Temperature [500 mb] (°C) Precipitation Total (mm)  \
209093                    -29.82                        0   
209094                     -29.7                        0   
209095                    -29.43                        0   
209096                    -29.65                        0   
209097                    -29.05                        0   

       Relative Humidity [2 m] (%)  ...  \
209093                    87.41894  ...   
209094                    89.29355  ...   
209095                   89.710396  ...   
209096                     89.9415  ...   
209097                    89.97886  ...   

       Soil Temperature [28-100 cm down] (°C)  \
209093                                  11.96   
209094                                  11.95   
209095                                  11.94   
209096                                  11.94   
209097                                  11.92   

       Soil Temperature [100-255 cm down] (°C)  \
209093                                   16.48   
209094                                   16.48   
209095                                   16.47   
209096                                   16.47   
209097                                   16.47   

       Soil Moisture [0-7 cm down] (m³/m³)  \
209093                                0.38   
209094                               0.379   
209095                               0.378   
209096                               0.379   
209097                               0.378   

       Soil Moisture [7-28 cm down] (m³/m³)  \
209093                                0.372   
209094                                0.372   
209095                                0.372   
209096                                0.373   
209097                                0.373   

       Soil Moisture [28-100 cm down] (m³/m³)  \
209093                                  0.276   
209094                                  0.276   
209095                                  0.276   
209096                                  0.278   
209097                                  0.278   

       Soil Moisture [100-255 cm down] (m³/m³)  \
209093                                   0.301   
209094                                   0.301   
209095                                   0.301   
209096                                   0.301   
209097                                   0.301   

       Soil Moisture Available To Plant [0-7 cm down] (fraction)  \
20

In [166]:
data.isnull().sum()

Timestamp                                                        0
Temperature [2 m elevation corrected] (°C)                       0
Growing Degree Days [2 m elevation corrected] (GDDc)             0
Temperature [900 mb] (°C)                                        0
Temperature [850 mb] (°C)                                        0
                                                                ..
Soil Moisture [100-255 cm down] (m³/m³)                          0
Soil Moisture Available To Plant [0-7 cm down] (fraction)        0
Soil Moisture Available To Plant [7-28 cm down] (fraction)       0
Soil Moisture Available To Plant [28-100 cm down] (fraction)     0
Soil Moisture Available To Plant [100-255 cm down] (fraction)    0
Length: 64, dtype: int64

Now that there are no more null values, we then have to seperate the Timestamp column into data and time

In [167]:
data['Timestamp'] = pd.to_datetime(data['Timestamp'])
data['Date'], data['Time'] = data['Timestamp'].dt.normalize(), data['Timestamp'].dt.time
data = data.drop('Timestamp', axis=1)

# Move the two new columns to the begining of the dataframe
data = data[['Date', 'Time'] + [col for col in data.columns if col not in ['Date', 'Time']]]

data.head()

Date      Time Temperature [2 m elevation corrected] (°C)  \
9  2000-01-01  00:00:00                                   1.550245   
10 2000-01-01  01:00:00                                   1.500245   
11 2000-01-01  02:00:00                                   1.470245   
12 2000-01-01  03:00:00                                   1.880245   
13 2000-01-01  04:00:00                                   2.310246   

   Growing Degree Days [2 m elevation corrected] (GDDc)  \
9                                                   0     
10                                                  0     
11                                                  0     
12                                                  0     
13                                                  0     

   Temperature [900 mb] (°C) Temperature [850 mb] (°C)  \
9                      -0.34                     -2.35   
10                      -0.3                     -2.63   
11                     -0.18                     -2.75   
12                     -0.15                     -2.77   
13                     -0.29                     -2.74   

   Temperature [800 mb] (°C) Temperature [700 mb] (°C)  \
9                      -4.99                     -8.75   
10                     -4.67                     -8.95   
11                     -4.52                     -8.63   
12                     -4.37                        -8   
13                     -4.44                     -7.59   

   Temperature [500 mb] (°C) Precipitation Total (mm)  ...  \
9                     -23.01                        0  ...   
10                    -22.71                        0  ...   
11                    -22.79                        0  ...   
12                    -22.85                        0  ...   
13                    -22.62                        0  ...   

   Soil Temperature [28-100 cm down] (°C)  \
9                                    3.85   
10                                   3.81   
11                                    3.8   
12                                    3.8   
13                                   3.79   

   Soil Temperature [100-255 cm down] (°C)  \
9                                     7.54   
10                                    7.63   
11                                    7.63   
12                                    7.63   
13                                    7.62   

   Soil Moisture [0-7 cm down] (m³/m³) Soil Moisture [7-28 cm down] (m³/m³)  \
9                                0.382                                0.385   
10                               0.384                                0.386   
11                               0.384                                0.385   
12                               0.384                                0.385   
13                               0.384                                0.385   

   Soil Moisture [28-100 cm down] (m³/m³)  \
9                                   0.416   
10                                  0.418   
11                                  0.418   
12                                  0.418   
13                                  0.417   

   Soil Moisture [100-255 cm down] (m³/m³)  \
9                                    0.395   
10                                   0.401   
11                                   0.401   
12                                   0.401   
13                                   0.401   

   Soil Moisture Available To Plant [0-7 cm down] (fraction)  \
9                                               0.231          
10                                              0.233          
11                                              0.233          
12                                              0.233          
13                                              0.233          

   Soil Moisture Available To Plant [7-28 cm down] (fraction)  \
9                                               0.234           
10                                              0.235           
11         

### 3. Explanatory Data Analysis

### 4. Feature Selection